torchtitan/components/quantization/mx_moe.py

```diff
@@ -0,0 +1,103 @@
+from torchtitan.protocols.model_converter import (
+    ModelConverter,
+    register_model_converter,
+)
+from torchtitan.tools.logging import logger
+from torchtitan.tools.utils import has_cuda_capability
+
+
+class MXMoEConverter(ModelConverter):
+    """Converts target 3D nn.Parameters of a model, representing 'experts',
+    to use MXFP8 scaled grouped GEMMs instead of a high precision grouped GEMMs."""
+
+    enabled: bool
+    filter_fqns: List[str]
+
+    def __init__(self, job_config: JobConfig, parallel_dims: ParallelDims):
+        # Ensure minimum torchao versions
+        if find_spec("torchao") is None:
+            raise ImportError(
+                "torchao is not installed. Please install it to use MXFP8 linear layers."
+            )
+        torchao_version = version("torchao")
+
+        # Require latest release or nightly builds for prototype features
+        is_nightly_build = torchao_version.startswith("0.14.0")
+        if not is_nightly_build:
+            raise ImportError(
+                f"torchao version {torchao_version} is too old, please install torchao nightly build and try again"
+            )
+
+        # Can be removed if we enable the emulated versions
+        assert has_cuda_capability(
+            10, 0
+        ), "MXFP8 is only supported on SM100 or architectures"
+
+        # Warn user if torch.compile is not enabled
+        model_compile_enabled = (
+            job_config.compile.enable and "model" in job_config.compile.components
+        )
+        if not model_compile_enabled:
+            logger.warning(
+                "torch.compile enablement is required for highest performance of MXFP8 dynamic quantization."
+            )
+
+        # For MoE training with mxfp8, token group sizes must be multiples of 32
+        self.moe_fqns = job_config.mx_moe.fqns
+        if self.moe_fqns:
+            logger.info(
+                f"Setting token group alignment size to {MXFP8_GROUP_ALIGNMENT_SIZE}"
+            )
+            set_token_group_alignment_size_m(MXFP8_GROUP_ALIGNMENT_SIZE)
+
+        self.enabled = True
+        logger.info("MXFP8 MoE training enabled")
+
+    def convert(self, model: nn.Module):
+        """
+        Mutates the model inplace replacing instances of nn.Parameter with ScaledGroupedMMTensor,
+        to perform dynamic MXFP8 quantization + scaled grouped GEMMs for the target MoE FQNs.
+        """
+        from torchao.prototype.moe_training.conversion_utils import (
+            MoEScalingType,
+            MoETrainingConfig,
+        )
+        from torchao.quantization.quant_api import quantize_
+
+        def moe_module_filter_fn(mod: nn.Module, cur_fqn: str) -> bool:
+            for target_fqn in self.moe_fqns:
+                if target_fqn in cur_fqn:
+                    return True
+            return False
+
+        config = MoETrainingConfig(scaling_type=MoEScalingType.MXFP8)
```

Comment:
I think the torchtitan feature name here should match the torchao one. If you want mxtraining, make the torchao workflow adhere to that too.